---
title: "Human integration"
author: "James Ashmore"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Motivation

## Setup

Setup chunk options:

```{r}
knitr::opts_chunk$set(
  autodep    = TRUE,
  cache      = TRUE,
  cache.lazy = FALSE,
  cache.path = "cache/03-human-integration.Rmd/",
  dev        = "png",
  error      = FALSE,
  message    = FALSE,
  warning    = FALSE,
  fig.align = "center",
  out.width = '100%'
)
```

Load Bioconductor packages:

```{r}
library(batchelor)
library(scater)
library(scran)
library(scuttle)
```

Load CRAN packages:

```{r}
library(ggforce)
library(ggrepel)
library(pheatmap)
```

## Data

### Mouse data

Read mouse experiment objects:

```{r}
sce <- list(
  NMP_EHF = readRDS("data/NMP_EHF.rds"),
  LPMP_EHF = readRDS("data/LPMP_EHF.rds"),
  LPMP_LHF = readRDS("data/LPMP_LHF.rds")
)
```

Define function to clean mouse experiment objects:

```{r}
cleanSCE <- function(x, sample = NA) {
  
  # Clear metadata
  metadata(x) <- list()
  
  # Remove NA
  x <- x[!is.na(rowData(x)$gene_id), ]

  # Replace rownames
  rownames(x) <- rowData(x)$gene_id

  # Clear rowData
  rowData(x) <- NULL

  # Replace rowRanges
  rowRanges(x) <- relist(GRanges(), PartitioningByEnd(integer(length(x)), names = rownames(x)))

  # Clear colnames
  colnames(x) <- NULL

  # Replace colData
  colData(x) <- DataFrame(
    Sample = sample,
    Cluster = paste(sample, x$cluster, sep = "_"),
    Origin = NA,
    sizeFactor = sizeFactors(x)
  )

  # Clear reducedDims
  reducedDims(x) <- NULL

  # Clear altExp
  altExp(x) <- NULL

  # Return object
  return(x)
  
}
```

Clean mouse experiment objects:

```{r}
sce <- mapply(cleanSCE, x = sce, sample = names(sce))
```

### Human data

Read human embryo experiment object:

```{r}
ref <- readRDS("data/HumanGastrulationData.rds")
```

Define function to clean human experiment object:

```{r}
cleanSCE <- function(x, sample = NA) {
  
  # Clear metadata
  metadata(x) <- list()
  
  # Remove NA
  x <- x[!is.na(rowData(x)$ID), ]

  # Replace rownames
  rownames(x) <- rowData(x)$ID

  # Clear rowData
  rowData(x) <- NULL

  # Replace rowRanges
  rowRanges(x) <- relist(GRanges(), PartitioningByEnd(integer(length(x)), names = rownames(x)))

  # Clear colnames
  colnames(x) <- NULL

  # Replace colData
  colData(x) <- DataFrame(
    Sample = sample,
    Cluster = x$Cluster,
    Origin = x$Origin,
    sizeFactor = sizeFactors(x)
  )

  # Clear reducedDims
  reducedDims(x) <- NULL

  # Clear altExp
  altExp(x) <- NULL

  # Return object
  return(x)
  
}
```

Clean human experiment object:

```{r}
ref <- cleanSCE(ref, sample = "Human")
```

## Integration

### Data processing

Read orthologous gene mapping table:

```{r}
tbl <- read.csv("data/hsapiens_mmusculus.csv")

tbl <- subset(tbl, mmusculus_homology_type == "ortholog_one2one")
```

Subset mapping table to measured genes:

```{r}
tbl <- subset(tbl, mmusculus_gene_id %in% Reduce(intersect, lapply(sce, rownames)))

tbl <- subset(tbl, hsapiens_gene_id %in% rownames(ref))
```

Subset experiment objects to common genes:

```{r}
sce <- lapply(sce, "[", i = tbl$mmusculus_gene_id)

ref <- ref[tbl$hsapiens_gene_id, ]
```

Replace mouse rowData slot with gene mapping table:

```{r}
sce <- lapply(sce, function(x) {
  
  df <- DataFrame(
    Human_ID = tbl$hsapiens_gene_id,
    Human_Symbol = tbl$hsapiens_gene_name,
    Mouse_ID = tbl$mmusculus_gene_id,
    Mouse_Symbol = tbl$mmusculus_gene_name,
    row.names = tbl$mmusculus_gene_id
  )
  
  df <- df[rownames(x), ]
  
  rowData(x) <- df

  rownames(x) <- rowData(x)$Human_ID
  
  return(x)
  
})
```  

Replace human rowData slot with gene mapping table:

```{r}
df <- DataFrame(
  Human_ID = tbl$hsapiens_gene_id,
  Human_Symbol = tbl$hsapiens_gene_name,
  Mouse_ID = tbl$mmusculus_gene_id,
  Mouse_Symbol = tbl$mmusculus_gene_name,
  row.names = tbl$hsapiens_gene_id
)

df <- df[rownames(ref), ]

rowData(ref) <- df

rownames(ref) <- rowData(ref)$Human_ID
```

Create list of all experiment data:

```{r}
all <- c("Human" = ref, sce)
```

Delete large objects and trigger garbage collection:

```{r}
rm(sce, ref, cleanSCE, tbl, df); gc()
```

### Variance modelling

Model the per-gene variance:

```{r}
dec <- lapply(all, modelGeneVar)
```

Combine variance decompositions:

```{r}
dec <- do.call(combineVar, dec)
```

### Feature selection

Identify highly variable genes:

```{r}
hvg <- getTopHVGs(dec, var.field = "bio", var.threshold = 0)
```

### Batch normalization

Perform scaling normalization:

```{r}
all <- multiBatchNorm(all)
```

### Batch integration

Correct experiments using nearest neighbours:

```{r}
set.seed(1000110100)

mnn <- correctExperiments(
  all,
  subset.row = hvg,
  correct.all = TRUE,
  PARAM = FastMnnParam(auto.merge = TRUE)
)
```

Delete large objects and trigger garbage collection:

```{r}
rm(all); gc()
```

Store combined variance model:

```{r}
metadata(mnn)$combineVar <- dec
```

Store highly variable genes:

```{r}
metadata(mnn)$getTopHVGs <- hvg
```

## Diagnostics

Check proportion of lost variance:

```{r}
var <- metadata(mnn)$merge.info$lost.var

col <- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = "Reds"))(100)

brk <- seq(0, max(var), length.out = 101)

pheatmap(var, color = col, display_numbers = TRUE, number_format = "%.2f", cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)
```

Cluster on the corrected PCA to obtain a partitioning of the cells:

```{r}
snn <- buildSNNGraph(mnn, type = "jaccard", use.dimred = "corrected")

com <- igraph::cluster_louvain(snn)
```

Tabulate number of cells from multiple batches within each cluster:

```{r}
tab <- table(Cluster = com$membership, Batch = mnn$batch)

col <- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = "Blues"))(100)

brk <- seq(0, max(tab), length.out = 101)

pheatmap(tab, color = col, display_numbers = TRUE, number_format = "%d", cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)
```

Perform UMAP on the corrected PCA matrix:

```{r}
set.seed(0001000010)

mnn <- runUMAP(mnn, dimred = "corrected", n_neighbors = 15, min_dist = 0.5) # same settings as reference analysis
```

Visualize the corrected PCA using a UMAP plot:

```{r fig.height=6, fig.width=8}
dat <- makePerCellDF(mnn, use.coldata = TRUE, use.dimred = "UMAP")

dat$Cluster[dat$Sample != "Human"] <- NA # ignore mouse clusters

HumanClusterColours <- c(
  "Epiblast" = "#3D7AB0",
  "Ectoderm" = "#4FC0CB",
  "Primitive Streak" = "#A6BEDF",
  "Nascent Mesoderm" = "#C67DB4",
  "Axial Mesoderm" = "#7B534D",
  "Emergent Mesoderm" = "#ABB973",
  "Endoderm" = "#F28941",
  "Advanced Mesoderm" = "#4F9E74",
  "YS Mesoderm" = "#FAB781",
  "Hemogenic Endothelial Progenitors" = "#6A59A6",
  "Erythrocytes" = "#C5433E"
)

ggplot(dat, aes(UMAP.1, UMAP.2, colour = Cluster)) + 
  geom_point(data = dat[, c("UMAP.1", "UMAP.2")], colour = "gainsboro") + 
  geom_point(size = 2) + 
  scale_colour_manual(values = HumanClusterColours, na.value = "#000000") + 
  labs(x = "UMAP 1", y = "UMAP 2") + 
  facet_wrap(~ batch) + 
  theme_no_axes(base.theme = theme_bw(base_size = 16)) + 
  theme(aspect.ratio = 1)
```

## Summary

### Output

Save integrated experiment object to disk:

```{r}
saveRDS(mnn, file = "output/03-human-integration.rds")
```

### Session

Print session information:

```{r}
sessionInfo()
```
